File: ts/src/rsa4096/jwk.ts
/**
 * Encodes an assembled key as a canonical Arweave JWK, and derives the
 * Arweave address from it -- exactly matching the real
 * arweave-core/src/keys/{address,keyfile}.ts logic (read from
 * AncientPantheon/constructors/Codex/packages/arweave-core during the
 * original research session): kty="RSA", e="AQAB", n decodes to exactly
 * 512 bytes, address = Base64URL(SHA-256(n)). Direct port of
 * RSA4096/jwk.go.
 *
 * Copyright (C) 2026 AncientHoldings GmbH. All rights reserved.
 */

import { sha256 } from '@noble/hashes/sha2.js';
import { bigIntToBytes } from './bigint-math.js';
import type { RSAKey } from './keyassembly.js';

/** Canonical 4096-bit modulus length in bytes. Matches RSA4096/jwk.go. */
export const ARWEAVE_MODULUS_BYTES = 512;

/** The canonical 9-field Arweave keyfile shape. Matches RSA4096/jwk.go's JWK. */
export interface JWK {
  kty: string;
  n: string;
  e: string;
  d: string;
  p: string;
  q: string;
  dp: string;
  dq: string;
  qi: string;
}

/** Unpadded base64url encoding, matching RFC 7518 (JWK) integer encoding. */
function base64url(bytes: Uint8Array): string {
  let base64: string;
  if (typeof Buffer !== 'undefined') {
    base64 = Buffer.from(bytes).toString('base64');
  } else {
    base64 = btoa(String.fromCharCode(...bytes));
  }
  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

/**
 * Encodes `key` as a canonical Arweave JWK. `n` is required to decode to
 * exactly {@link ARWEAVE_MODULUS_BYTES} bytes (guaranteed by construction
 * -- see RSA4096/jwk.go's ToJWK for the arithmetic proof, mirrored here).
 * Matches RSA4096/jwk.go's ToJWK exactly.
 */
export function toJWK(key: RSAKey): JWK {
  const nBytes = bigIntToBytes(key.n);
  if (nBytes.length !== ARWEAVE_MODULUS_BYTES) {
    throw new Error('toJWK: n is not exactly 512 bytes -- bit-fixing invariant violated upstream');
  }

  return {
    kty: 'RSA',
    n: base64url(nBytes),
    e: base64url(bigIntToBytes(key.e)),
    d: base64url(bigIntToBytes(key.d)),
    p: base64url(bigIntToBytes(key.p)),
    q: base64url(bigIntToBytes(key.q)),
    dp: base64url(bigIntToBytes(key.dp)),
    dq: base64url(bigIntToBytes(key.dq)),
    qi: base64url(bigIntToBytes(key.qi)),
  };
}

/**
 * Derives the Arweave address Base64URL(SHA-256(n)) directly from the
 * modulus, matching arweave-core's addressOf() exactly. Independently
 * re-validates n's decoded length itself (self-contained, doesn't depend
 * on toJWK having run first). Matches RSA4096/jwk.go's AddressOf exactly.
 */
export function addressOf(n: bigint): string {
  const nBytes = bigIntToBytes(n);
  if (nBytes.length !== ARWEAVE_MODULUS_BYTES) {
    throw new Error('addressOf: n is not exactly 512 bytes');
  }
  return base64url(sha256(nBytes));
}

File: research/rsa4096-poc/jwk.go
// This file implements the local (zero-network, zero-funds) half of STEP 8:
// encoding the assembled key as a canonical Arweave JWK, and deriving the
// Arweave address from it -- exactly matching the real
// arweave-core/src/keys/{address,keyfile}.ts logic read from
// AncientPantheon/constructors/Codex/packages/arweave-core during this
// session (kty="RSA", e="AQAB", n decodes to exactly 512 bytes, address =
// Base64URL(SHA-256(n))).
package main

import (
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"math/big"
)

// arweaveModulusBytes is the canonical 4096-bit modulus length in bytes,
// matching arweave-core's MODULUS_BYTES constant exactly.
const arweaveModulusBytes = 512

// JWK is the canonical 9-field Arweave keyfile shape (mirrors arweave-core's
// ArweaveJwk type exactly: kty + 8 base64url string fields).
type JWK struct {
	Kty string `json:"kty"`
	N   string `json:"n"`
	E   string `json:"e"`
	D   string `json:"d"`
	P   string `json:"p"`
	Q   string `json:"q"`
	Dp  string `json:"dp"`
	Dq  string `json:"dq"`
	Qi  string `json:"qi"`
}

// b64url encodes b as unpadded base64url, matching RFC 7518 (JWK) integer
// encoding and arweave-core's expected alphabet.
func b64url(b []byte) string {
	return base64.RawURLEncoding.EncodeToString(b)
}

// ToJWK encodes k as a canonical Arweave JWK. n is required to be exactly
// arweaveModulusBytes bytes (guaranteed by construction: both p and q have
// their top two bits forced to 1 by GenerateCandidate, so their product
// always has bit-length exactly 4096 -- see the comment in this function for
// the arithmetic proof). Every other field is encoded as its natural
// minimal big-endian byte representation, with no artificial padding --
// identical to how Node/browser WebCrypto export JWKs, and to what
// arweave-core's importKeyfile validates.
func (k *RSAKey) ToJWK() (*JWK, error) {
	nBytes := k.N.Bytes()
	if len(nBytes) != arweaveModulusBytes {
		// Proof this should never happen: p, q each have bit-length exactly
		// 2048 (top two bits forced to 1 by GenerateCandidate), so each lies
		// in [0.75 * 2^2048, 2^2048). Their product n = p*q therefore lies in
		// [0.5625 * 2^4096, 2^4096). Since 0.5625 > 0.5, n's top bit is
		// always 1, so n's bit-length is always exactly 4096 = 512 bytes.
		return nil, errors.New("ToJWK: n is not exactly 512 bytes -- bit-fixing invariant violated upstream")
	}

	return &JWK{
		Kty: "RSA",
		N:   b64url(nBytes),
		E:   b64url(k.E.Bytes()),
		D:   b64url(k.D.Bytes()),
		P:   b64url(k.P.Bytes()),
		Q:   b64url(k.Q.Bytes()),
		Dp:  b64url(k.Dp.Bytes()),
		Dq:  b64url(k.Dq.Bytes()),
		Qi:  b64url(k.Qi.Bytes()),
	}, nil
}

// AddressOf derives the Arweave address Base64URL(SHA-256(n)) directly from
// the modulus, matching arweave-core's addressOf() exactly (verified against
// its actual source this session). Independently re-validates n's decoded
// length itself (does NOT depend on ToJWK having run first, or on any
// particular call order) -- this function has the same "reject a
// wrong-length modulus" guard arweave-core's addressOf() has, self-contained.
func AddressOf(n *big.Int) (string, error) {
	nBytes := n.Bytes()
	if len(nBytes) != arweaveModulusBytes {
		return "", errors.New("AddressOf: n is not exactly 512 bytes")
	}
	digest := sha256.Sum256(nBytes)
	return b64url(digest[:]), nil
}
